### ISSUE TITLE
Country field doesn't render

### ISSUE THREAD

**Issue by MegaphoneJon** [eileenmcnaughton/civicrm_entity#198]
On CiviCRM Address entities, the State/Province renders but the Country doesn't, even though these fields should be roughly identical in function.

As a sidenote, I tried troubleshooting this, but I don't know enough about how to force a call to `Drupal\civicrm_entity\Entity\FieldDefinitionProvider::getBaseFieldDefinition`.  I assume this is cached somehow, since it's not triggered when modifying a View; if you can give me a tip on where to look I can start fixing some of my own problems and sending PRs back.

**Comment by KarinG** [comment]
Hi @MegaphoneJon - I hope you are well. 

I just wanted to add a note that I've reproduced this issue - and that we'll be prioritizing it.
 
![image](https://user-images.githubusercontent.com/5340555/78461655-367ccd80-7688-11ea-911c-1f818979250b.png)

**Comment by jackrabbithanna** [comment]
Is this still an issue? I'm not able to replicate this with current version. 
I tried a View with a base table of Address and a second view with a base table of contact that adds a relationship to the Address. 
On both I can see the country field produces values.

**Comment by jackrabbithanna** [comment]
FYI .. testing on a installation where the Civi tables are in a separate DB, and similar database integration in the settings.php that we had to use for Drupal 7.

I wonder if that is making a difference for me. Is this issue unique to standard D8 install with civi tables in Drupal db and no prefix information in settings.php

**Comment by KarinG** [comment]
@jackrabbithanna - I can confirm I’m seeing this on s standard D8 install (w/ databases in the same table).

**Comment by jackrabbithanna** [comment]
Interesting. Ok, definitely a clue that it does work how I have Civi setup.

**Comment by jackrabbithanna** [comment]
I've confirmed this issue exists when Civi tables are in the Drupal DB

**Comment by dsnopek** [comment]
I started poking around at this.

The difference between the 'State/Province' and 'Country' fields, is that 'State/Province' is implemented as a 'list_integer' field which looks up the values via `civicrm_api3($entity, 'getoptions', ...)`, whereas the 'Country' filed is implemented as an 'entity_reference' field to 'civicrm_country' entities.

The reason for that difference, is that `\Drupal\civicrm_entity\SupportedEntities::getInfo()` has an entry for 'civicrm_country' but no entry for 'civicrm_state_province'.

I don't actually recall how well developed the entity reference support is in civicrm_entity, but it appears to just not work in this case for some reason...

**Comment by dsnopek** [comment]
Do we really even need 'civicrm_country' to be entities? Like, we're not going to create, update or delete them. We're (probably?) not going to add configurable fields to them. Maybe the solution here is just to remove them from `\Drupal\civicrm_entity\SupportedEntities::getInfo()`? Of course, we should still fix whatever's wrong with entity references, but I'm not sure we need that to fix this bug...

**Comment by dsnopek** [comment]
Here's a PR that does that:

https://github.com/eileenmcnaughton/civicrm_entity/pull/211

Not sure it's the right way to go, so feedback welcome!

**Comment by jackrabbithanna** [comment]
I did a little testing .. and if I enable the "Country" entity type here: admin/structure/civicrm-entity/settings
Then the issue is fixed.
Keeping it as an entity type, gives automatic Views integration for the country table, which is a good thing.
This does raise on issue, if "Address" entity type is enabled, then Country should also be enabled. So if we keep the entity type, should there be some validation on the settings form?

Strange that it also doesn't manifest if Civi tables are not in the same DB as Drupal tables.

**Comment by KarinG** [comment]
Confirming that enabling Country as an Entity Type -> does show Country (and more! bonus info!) -> 

![image](https://user-images.githubusercontent.com/5340555/80644170-74baa200-8a26-11ea-8745-75bc10006bce.png)

But I agree with @dsnopek that I would not want to manage/create e.g. Countries -> therefore I never even thought about enabling it as an Entity under Settings. 

Let me try see what this looks like when Country is not an Entity ( PR #211 )

Ok - that worked and I think this solution makes more sense ->

![image](https://user-images.githubusercontent.com/5340555/80644903-a6803880-8a27-11ea-93f9-d7a8e87342fb.png)

**Comment by jackrabbithanna** [comment]
@Karin .. your View would look the same if you choose the "Entity Label" instead of "Rendered Entity" formatter for the country field.

One of the advantages of it being an entity type, is to be able to manage displays for it, and when choosing "Rendered Entity" be able to choose a view mode. 

I'd prefer if Country is an entity type, mainly to have Views integration for that table and a Views Relationship from Address table to it. 

Alternatively, if the field had a formatter to choose which property to display, like ISO code, country_code, or full name, that would suffice for those people who want options for what to display for the country field in a View.

People could include it back in SupportedEntities.php if they choose. One day I'd like to do a hook or plugin system so that developers can add entity types or alter the permissions settings. We used that frequently in the D7 version.

Since we've discovered an easy fix, I lean toward not merging it, although it is non-obvious what to do .. bad UX.

All that being said, in the spirit of cooperation, better UX,  and not spending a bunch of time with complex settings form validation or any of the other options, and later doing improvements, I won't fuss if the PR does get merged.

**Comment by KarinG** [comment]
I was curious to see what Country would show as a Rendered Entity (what more there was to it) - so had to try that. 

I really feel having to make the Country an Entity (which is not intuitive as who would be creating/updating countries) is like a work around. No one will know to do that -> so my vote would be that we merge this - it seems more intuitive. It just works.

**Comment by dsnopek** [comment]
> I did a little testing .. and if I enable the "Country" entity type here: admin/structure/civicrm-entity/settings
> Then the issue is fixed.

Eeep, this is another bug! The goal that Matt and I had was for all entity types to always fully exist, but for those checkboxes to just enable/disable whether or not they were exposed in the UI, to avoid _exactly_ this kind of problem. This sort of thing could come up not just in Views, but on entity edit pages (for the Entity Reference widget to work) and with the Drupal Rest API or Rules or whatever.

So, somewhere, the checkbox is actually doing more than it should...

**Comment by dsnopek** [comment]
I poked around the bug with disabled entities messing up the entity references...

It looks like the code is more-or-less written as intended: `civicrm_entity_entity_type_build()` is always defining all entities (in SupportedEntities at least), but for ones that are "enabled", it'll additionally add the forms, routes and field UI. I played around with moving some of that stuff up to the default definition to see if I could find which piece is needed to get entity references fully working, and I wasn't able to find it.

Digging a little deeper, I found a reference in `CivicrmEntityAccessHandler::access()` - it's denying access to all entities that aren't enabled in the UI - that's the culprit! Here's a new PR that fixes this issue:

https://github.com/eileenmcnaughton/civicrm_entity/pull/212

That PR also fixes issue #198, but still leaves civicrm_country exposed as an entity :-)
